src/lib/yahooFinanceTypes.ts

```diff
@@ -0,0 +1,138 @@
+import { Type } from "@sinclair/typebox";
+import { FormatRegistry } from "@sinclair/typebox";
+import { isDate, isDateTime } from "./datetime";
+
+FormatRegistry.Set("date", isDate);
+FormatRegistry.Set("date-time", isDateTime);
+
+// Strictly must be empty
+const EmptyObjectCoerceToNull = Type.Transform(
+  Type.Object({}, { maxProperties: 0 })
+)
+  .Decode(() => null)
+  .Encode(() => ({}));
+
+// Technically this will also contain a string 'fmt' key but we don't care because we don't use it
+const RawNumber = Type.Transform(
+  Type.Object({
+    raw: Type.Number(),
+  })
+)
+  .Decode((v) => v.raw)
+  .Encode((v) => ({ raw: v }));
+
+const TwoNumberRangeString = Type.Transform(
+  Type.RegExp(/^(-?\d+(?:\.\d+)?) - (-?\d+(?:\.\d+)?)$/g)
+)
+  .Decode((value) => {
+    // Split the two numbers allowing for negatives on either side
+    const validatedNumbers = value.match(/-?\d+(?:\.\d+)?/g);
+
+    if (!validatedNumbers) {
+      throw new Error(`Unable to decode number range from: ${value}`);
+    }
+
+    const [low, high] = validatedNumbers.map((number) => parseFloat(number));
+
+    if (isNaN(low) || isNaN(high)) {
+      throw new Error(
+        `Unable to decode number range from: ${value}. Decoded value for low is: ${low}, decoded value for high is: ${high}`
+      );
+    }
+    return { low, high };
+  })
+  .Encode(({ low, high }) => `${low.toFixed(2)} - ${high.toFixed(2)}`);
+
+const TwoNumberRange = Type.Object({
+  low: Type.Number(),
+  high: Type.Number(),
+});
+
+const EpochTimestamp = Type.Transform(Type.Number())
+  .Decode((v) => new Date(v * 1000))
+  .Encode((v) => +v / 1000);
+
+const RawDateObject = Type.Transform(
+  Type.Object({
+    raw: EpochTimestamp,
+  })
+)
+  .Decode((v) => v.raw)
+  .Encode((v) => ({
+    raw: v,
+  }));
+
+const ISOStringDate = Type.Transform(
+  Type.Union([
+    Type.String({ format: "date" }),
+    Type.String({ format: "date-time" }),
+  ])
+)
+  .Decode((v) => new Date(v))
+  .Encode((v) => v.toISOString());
+
+export const YahooFinanceDate = Type.Union([
+  Type.Date(),
+  EpochTimestamp,
+  RawDateObject,
+  ISOStringDate,
+]);
+
+/**
+ * Validates and decodes all nullable date representations produced by Yahoo
+ * e.g. accepted inputs include:
+ * - 1612313997
+ * - { raw: 1612313997 }
+ * - "2024-02-29"
+ * - "2024-05-04T13:24:41.100Z"
+ * - {} (coerces to null)
+ */
+// Taking suggestions on the naming of this type, it's not *just* nullable it also accepts {} as per the unit tests
```

Comment:
Based on the existing test functionality it seems like nullable inputs are also required to coerce `{}` to `null`.

That's fine, but it does make the name of the type slightly misleading. It's not *just* nullable, it can also be an empty object. 

Hardest problem in CS 🙃

Reply:
Haha, hardest problem indeed.  Although I'm perfectly comfortable with the name, which reflects the intention rather than all the nitty-gritty details which *users* of the library won't care about.  For the devs you've made it clear in the comments.  At the end of the day, it's a function that given a date/null in any of the many varieties Yahoo provides, we'll get back a real `Date` or `null`.  If anyone else is reading this and feels differently, do let us know!

Reply:
Coolio, I've taken out the comments asking for suggestions but left the rest as is.